packages/core/src/directory.ts

```diff
@@ -0,0 +1,115 @@
+}
+
+export interface DirectoryProvider {
+    subscribe(sub: DirectorySubscriber): void;
+}
+
+export class DirectorySync implements DirectorySubscriber {
+    constructor(public readonly server: Server, providers: DirectoryProvider[] = []) {
+        for (const provider of providers) {
+            provider.subscribe(this);
+        }
+    }
+
+    async userCreated(user: DirectoryUser, orgId: string) {
+        const org = (orgId && (await this.server.storage.get(Org, orgId))) || null;
+        if (org && org.directory.syncProvider === "scim" && org.directory.syncMembers) {
+            const memberExists = org.members.some((member) => member.directoryExternalId === user.externalId);
+
+            if (memberExists) {
+                return;
+            }
+
+            org.members.push(
+                new OrgMember({
+                    directoryExternalId: user.externalId,
+                    name: user.name,
+                    email: user.email,
+                    status: OrgMemberStatus.Provisioned,
+                    updated: new Date(),
+                })
+            );
+
+            await this.server.updateMetaData(org);
+            await this.server.storage.save(org);
+        }
+    }
+
+    async userUpdated(user: DirectoryUser, orgId: string): Promise<void> {
+        const org = (orgId && (await this.server.storage.get(Org, orgId))) || null;
+        if (org && org.directory.syncProvider === "scim" && org.directory.syncMembers) {
+            const existingUser = org.members.find((member) => member.directoryExternalId === user.externalId);
+
+            if (!existingUser) {
+                return;
+            }
+
+            existingUser.name = user.name;
+            existingUser.updated = new Date();
+
+            await this.server.updateMetaData(org);
+            await this.server.storage.save(org);
+        }
+    }
+    async userDeleted(user: DirectoryUser, orgId: string): Promise<void> {
+        const org = (orgId && (await this.server.storage.get(Org, orgId))) || null;
+        if (org && org.directory.syncProvider === "scim" && org.directory.syncMembers) {
+            const existingUser = org.members.find((member) => member.directoryExternalId === user.externalId);
+
+            if (!existingUser) {
+                return;
+            }
+
+            await org.removeMember(existingUser, false);
+
+            // Remove any existing invites
+            const existingInvite = org.invites.find((invite) => invite.email === existingUser.email);
+            if (existingInvite) {
+                org.removeInvite(existingInvite);
+
+                try {
+                    const auth = await this.server.getAuth(existingInvite.email);
+                    auth.invites = auth.invites.filter((invite) => invite.id !== existingInvite.id);
+                    await this.server.storage.save(auth);
+                } catch (_error) {
+                    // Ignore
+                }
```

Comment:
@MaKleSoft I had to create and expose the `getAuth` (a simplified version) to allow to do this. I think this is enough now? It seemed harder to expose/inject the controller here, though perhaps better?

Reply:
We may want to think of a cleaner/better way to do this in the future but for now this should be fine.